File: Java/Beginner/A_Sorting/Boj_24060.java
package Beginner.A_Sorting;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Boj_24060 {
    static StringTokenizer st;
    static int[] arr;
    static int ans = -1;
    static int cnt = 0;
    static int n, k;

    private static void merging(int left, int mid, int right) {
        int[] tmp = new int[right + 2];

        int i = left;
        int j = mid + 1;
        int t = left;

        while (i <= mid && j <= right) {
            if (arr[i] <= arr[j]) tmp[t++] = arr[i++];
            else tmp[t++] = arr[j++];
        }

        while (i <= mid)
            tmp[t++] = arr[i++];

        while (j <= right)
            tmp[t++] = arr[j++];

        for (int a = left; a <= right; a++) {
            if (++cnt == k) {
                ans = tmp[a];
                break;
            }
        }
    }

    private static void merge_sort(int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            merge_sort(left, mid);
            merge_sort(mid + 1, right);
            merging(left, mid, right);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br =
                new BufferedReader(new InputStreamReader(System.in));

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        arr = new int[n];
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++)
            arr[i] = Integer.parseInt(st.nextToken());

        merge_sort(0, n - 1);

        System.out.println(ans);
        br.close();
    }
}

File: Kotlin/src/main/kotlin/beginner/a_Sorting/boj_24060.kt
package beginner.a_Sorting

import java.io.BufferedReader
import java.io.InputStreamReader

var cnt = 0
var ans = -1
lateinit var tmp: IntArray

fun merging(arr: IntArray, left: Int, mid: Int, right: Int, k: Int) {
    var (i, j, t) = intArrayOf(left, mid + 1, left)

    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) tmp[t++] = arr[i++]
        else tmp[t++] = arr[j++]
    }

    while (i <= mid) tmp[t++] = arr[i++]
    while (j <= right) tmp[t++] = arr[j++]

    for (a in left..right) {
        if (++cnt == k) {
            ans = tmp[a]
            return
        }
        arr[a] = tmp[a]
    }
}

fun merge(arr: IntArray, left: Int, right: Int, k: Int) {
    if (left < right) {
        val mid = (left + right) / 2
        merge(arr, left, mid, k)
        merge(arr, mid + 1, right, k)
        merging(arr, left, mid, right, k)
    }
}

fun boj_24060() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (_, k) = readLine()
        .split(" ")
        .map { it.toInt() }

    val arr: IntArray = readLine()
        .split(" ")
        .map{ it.toInt() }
        .toIntArray()

    tmp = IntArray(arr.size)
    merge(arr, 0, arr.lastIndex, k)

    print(ans)
}